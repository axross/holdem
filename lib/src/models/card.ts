/**
 *
 */
export type Card = number;

/**
 *
 */
export enum Rank {
  Ace = 0,
  King = 1,
  Queen = 2,
  Jack = 3,
  Ten = 4,
  Nine = 5,
  Eight = 6,
  Seven = 7,
  Six = 8,
  Five = 9,
  Four = 10,
  Three = 11,
  Deuce = 12,
}

export enum Suit {
  Spade = 0,
  Heart = 1,
  Diamond = 2,
  Club = 3,
}

/**
 *
 */
export function stringToCard(value: string): Card {
  return 2 ** (ranks.indexOf(value[0]!) + suits.indexOf(value[1]!) * 13);
}

export function cardToString(value: Card): string {
  return `${ranks[getRank(value)]}${suits[getSuit(value)]}`;
}

export function getRank(card: Card): Rank {
  return Math.log2(card) % 13;
}

export function getSuit(card: Card): Suit {
  return ~~(Math.log2(card) / 13);
}

const ranks = ["A", "K", "Q", "J", "T", "9", "8", "7", "6", "5", "4", "3", "2"];
const suits = ["s", "h", "d", "c"];
