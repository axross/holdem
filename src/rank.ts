/**
 * An enum type that expresses a rank part of a card.
 */
export enum Rank {
  Ace = 0,
  Deuce = 1,
  Trey = 2,
  Four = 3,
  Five = 4,
  Six = 5,
  Seven = 6,
  Eight = 7,
  Nine = 8,
  Ten = 9,
  Jack = 10,
  Queen = 11,
  King = 12,
}

/**
 * A string expression of a rank. It's always 1-character length. You can combine this with a SuitChar to build a CardString.
 *
 * @example
 * ```
 * const rankChar: RankChar = "A";
 * const suitChar: SuitChar = "s";
 * const cardString: CardString = `${rankChar}${suitChar}`;
 * ```
 */
export type RankChar =
  | "A"
  | "2"
  | "3"
  | "4"
  | "5"
  | "6"
  | "7"
  | "8"
  | "9"
  | "T"
  | "J"
  | "Q"
  | "K";

/**
 * A utility function set for ranks.
 */
export const RankUtils = Object.freeze({
  /**
   * Parses a char (= 1-charactor-length string) into a Rank.
   *
   * @example
   * ```ts
   * RankUtils.parse("A") === Rank.Ace;
   * RankUtils.parse("T") === Rank.Ten;
   * RankUtils.parse("5") === Rank.Five;
   * ```
   *
   * Only `"A"`, `"K"`, `"Q"`, `"J"`, `"T"`, `"9"`, `"8"`, `"7"`, `"6"`, `"5"`, `"4"`, `"3"` or `"2"` is acceptable.
   *
   * @example
   * ```ts
   * RankUtils.parse("a");  // => Error: "a" is not a valid string value for RankUtils.parse().
   * ```
   *
   * For consistency reason, it needs to be a string even for number-based ranks (e.g. `"4"`).
   *
   * @example
   * ```ts
   * RankUtils.parse("4") === Rank.Four;
   * RankUtils.parse(4);  // Error: 4 is not a valid string value for RankUtils.parse().
   * ```
   */
  parse(char: string): Rank {
    switch (char) {
      case "A":
        return Rank.Ace;
      case "K":
        return Rank.King;
      case "Q":
        return Rank.Queen;
      case "J":
        return Rank.Jack;
      case "T":
        return Rank.Ten;
      case "9":
        return Rank.Nine;
      case "8":
        return Rank.Eight;
      case "7":
        return Rank.Seven;
      case "6":
        return Rank.Six;
      case "5":
        return Rank.Five;
      case "4":
        return Rank.Four;
      case "3":
        return Rank.Trey;
      case "2":
        return Rank.Deuce;
      default:
        throw new Error(
          `${JSON.stringify(char)} is not a valid value for RankUtils.parse().`
        );
    }
  },

  /**
   * Returns power of a Rank.
   *
   * @example
   * ```ts
   * RankUtils.power(Rank.Ace) > RankUtils.power(Rank.King);    // => true
   * RankUtils.power(Rank.Ace) > RankUtils.power(Rank.Deuce);   // => true
   * RankUtils.power(Rank.Trey) > RankUtils.power(Rank.Deuce);  // => true
   * ```
   */
  power(rank: Rank): number {
    switch (rank) {
      case Rank.Ace:
        return 12;
      case Rank.King:
        return 11;
      case Rank.Queen:
        return 10;
      case Rank.Jack:
        return 9;
      case Rank.Ten:
        return 8;
      case Rank.Nine:
        return 7;
      case Rank.Eight:
        return 6;
      case Rank.Seven:
        return 5;
      case Rank.Six:
        return 4;
      case Rank.Five:
        return 3;
      case Rank.Four:
        return 2;
      case Rank.Trey:
        return 1;
      case Rank.Deuce:
        return 0;
    }
  },

  /**
   * Compares two ranks in index order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * RankUtils.compare(Rank.Ace, Rank.Deuce);  // => negative integer
   * RankUtils.compare(Rank.Deuce, Rank.Ace);  // => positive integer
   * RankUtils.compare(Rank.Ace, Rank.Ace);  // => 0
   * ```
   */
  compare(a: Rank, b: Rank): number {
    return a - b;
  },

  /**
   * Compares two ranks in power order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * RankUtils.compare(Rank.Ace, Rank.King);  // => negative integer
   * RankUtils.compare(Rank.King, Rank.Ace);  // => positive integer
   * RankUtils.compare(Rank.Ace, Rank.Ace);  // => 0
   * RankUtils.compare(Rank.Ace, Rank.Deuce);  // => positive integer
   * ```
   */
  comparePower(a: Rank, b: Rank): number {
    return RankUtils.power(b) - RankUtils.power(a);
  },

  /**
   * Stringifies a Rank.
   *
   * @example
   * ```ts
   * RankUtils.format(Rank.Ace) === "A";
   * RankUtils.format(Rank.Ten) === "T";
   * RankUtils.format(Rank.Five) === "5";
   * ```
   *
   * You can utilize this function to build a CardString.
   *
   * @example
   * ```ts
   * const aceChar = RankUtils.format(Rank.Ace);
   * const spadeChar = RankUtils.format(Suit.Spade);
   * const aceOfSpade: CardString = `${aceChar}${spadeChar}`;
   * ```
   */
  format(rank: Rank): RankChar {
    switch (rank) {
      case Rank.Ace:
        return "A";
      case Rank.King:
        return "K";
      case Rank.Queen:
        return "Q";
      case Rank.Jack:
        return "J";
      case Rank.Ten:
        return "T";
      case Rank.Nine:
        return "9";
      case Rank.Eight:
        return "8";
      case Rank.Seven:
        return "7";
      case Rank.Six:
        return "6";
      case Rank.Five:
        return "5";
      case Rank.Four:
        return "4";
      case Rank.Trey:
        return "3";
      case Rank.Deuce:
        return "2";
    }
  },
});
