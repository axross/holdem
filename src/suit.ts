/**
 * An enum type that expresses a suit part of a card.
 */
export enum Suit {
  Spade = 0,
  Heart = 1,
  Diamond = 2,
  Club = 3,
}

/**
 * A string expression of a suit. It's always 1-character length. You can combine this with a RankChar to build a CardString.
 *
 * @example
 * ```
 * const rankChar: RankChar = "A";
 * const suitChar: SuitChar = "s";
 * const cardString: CardString = `${rankChar}${suitChar}`;
 * ```
 */
export type SuitChar = "s" | "h" | "d" | "c";

/**
 * A utility function set for Suits.
 */
export const SuitUtils = Object.freeze({
  /**
   * Parses a char (= 1-charactor-length string) into a Suit.
   *
   * @example
   * ```ts
   * SuitUtils.parse("s") === Suit.Spade;
   * SuitUtils.parse("h") === Suit.Heart;
   * SuitUtils.parse("d") === Suit.Diamond;
   * SuitUtils.parse("c") === Suit.Club;
   * ```
   *
   * Only `"s"`, `"h"`, `"d"`, or `"c"` is acceptable.
   *
   * @example
   * ```ts
   * SuitUtils.parse("S");  // => Error: "S" is not a valid string value for SuitUtils.parse().
   * ```
   */
  parse(char: string): Suit {
    switch (char) {
      case "s":
        return Suit.Spade;
      case "h":
        return Suit.Heart;
      case "d":
        return Suit.Diamond;
      case "c":
        return Suit.Club;
      default:
        throw new Error(
          `${JSON.stringify(char)} is not a valid value for SuitUtils.parse().`
        );
    }
  },

  /**
   * Compares two ranks in index order and returns integer compatible with Array#sort().
   *
   * It results in the ordinal order that how it should usually be in poker.
   *
   * @example
   * ```
   * SuitUtils.compare(Suit.Spade, Suit.Heart);  // => negative integer
   * SuitUtils.compare(Suit.Club, Suit.Diamond);  // => positive integer
   * SuitUtils.compare(Suit.Heart, Suit.Heart);  // => 0
   * ```
   */
  compare(a: Suit, b: Suit): number {
    return a - b;
  },

  /**
   * Stringifies a Suit.
   *
   * @example
   * ```ts
   * SuitUtils.format(Suit.Spade) === "s";
   * SuitUtils.format(Suit.Heart) === "h";
   * SuitUtils.format(Suit.Diamond) === "d";
   * SuitUtils.format(Suit.Club) === "c";
   * ```
   *
   * You can utilize this function to build a CardString.
   *
   * @example
   * ```ts
   * const aceChar = RankUtils.format(Rank.Ace);
   * const spadeChar = RankUtils.format(Suit.Spade);
   * const aceOfSpade: CardString = `${aceChar}${spadeChar}`;
   * ```
   */
  format(suit: Suit): SuitChar {
    switch (suit) {
      case Suit.Spade:
        return "s";
      case Suit.Heart:
        return "h";
      case Suit.Diamond:
        return "d";
      case Suit.Club:
        return "c";
    }
  },
});
