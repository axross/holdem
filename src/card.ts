import { Rank, RankChar, RankUtils } from "./rank";
import { Suit, SuitChar, SuitUtils } from "./suit";

/**
 * An integer value that expresses a card. This value is always `2^0 <= n <= 2^51`.
 *
 * @example
 * ```
 * 0b1 as Card;      // ace of spade
 * 0b10 as Card;     // deuce of spade
 * 2 ** 51 as Card;  // king of club
 * ```
 */
export type Card = number & {
  __CardBrand: never;
};

/**
 * A string expression of a card. It's always 2-character length.
 *
 * @example
 * ```
 * "As" as CardString;  // ace of spade
 * "2d" as CardString;  // deuce of diamond
 * "Qh" as CardString;  // queen of heart
 * "Tc" as CardString;  // ten of club
 * ```
 */
export type CardString = `${RankChar}${SuitChar}`;

/**
 * A utility function set for Cards.
 */
export const CardUtils = Object.freeze({
  /**
   * Creates a Card from a given pair of Rank and Suit.
   */
  create(rank: Rank, suit: Suit): Card {
    return (2 ** (rank + suit * 13)) as Card;
  },

  /**
   * Parses a string into a Card.
   *
   * @example
   * ```ts
   * CardUtils.parse("As") === CardUtils.create(Rank.Ace, Suit.Spade);
   * CardUtils.parse("2s") === CardUtils.create(Rank.Deuce, Suit.Spade);
   * CardUtils.parse("Kc") === CardUtils.create(Rank.King, Suit.Club);
   * ```
   */
  parse(string: string): Card {
    if (!/^[A23456789TJQK][shdc]$/.test(string)) {
      throw new TypeError(
        `"${string}" is not a valid string for CardUtils.parse().`
      );
    }

    return CardUtils.create(
      RankUtils.parse(string[0] as RankChar),
      SuitUtils.parse(string[1] as SuitChar)
    );
  },

  /**
   * Creates a randomly-chosen Card.
   *
   * @example
   * ```ts
   * CardUtils.random();  => Card (at random)
   * ```
   */
  random(): Card {
    return (2 ** Math.floor(Math.random() * 52)) as Card;
  },

  /**
   * Extracts the Rank part of a Card.
   *
   * @example
   * ```
   * CardUtils.rankOf(CardUtils.parse("As")) === Rank.Ace;
   * CardUtils.rankOf(CardUtils.parse("3h")) === Rank.Trey;
   * CardUtils.rankOf(CardUtils.parse("Td")) === Rank.Ten;
   * CardUtils.rankOf(CardUtils.parse("6c")) === Rank.Six;
   * ```
   */
  rankOf(card: Card): Rank {
    return Math.log2(card) % 13;
  },

  /**
   * Extracts the Suit part of a Card.
   *
   * @example
   * ```
   * CardUtils.suitOf(CardUtils.parse("As")) === Suit.Spade;
   * CardUtils.suitOf(CardUtils.parse("3h")) === Suit.Heart;
   * CardUtils.suitOf(CardUtils.parse("Td")) === Suit.Diamond;
   * CardUtils.suitOf(CardUtils.parse("6c")) === Suit.Club;
   * ```
   */
  suitOf(card: Card): Suit {
    return ~~(Math.log2(card) / 13);
  },

  /**
   * Compares two cards in index order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * CardUtils.compare(Card(Rank.Ace, Suit.Spade), Card(Rank.Ace, Suit.Diamond));  // => negative integer
   * CardUtils.compare(Card(Rank.Ace, Suit.Diamond), Card(Rank.Six, Suit.Heart));  // => positive integer
   * CardUtils.compare(Card(Rank.Ace, Suit.Club), Card(Rank.Ace, Suit.Club));  // => 0
   * ```
   */
  compare(a: Card, b: Card): number {
    return (
      SuitUtils.compare(CardUtils.suitOf(a), CardUtils.suitOf(b)) * 13 +
      RankUtils.compare(CardUtils.rankOf(a), CardUtils.rankOf(b))
    );
  },

  /**
   * Compares two ranks in power order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * CardUtils.comparePower(Card(Rank.Ace, Suit.Spade), Card(Rank.Ace, Suit.Diamond));  // => negative integer
   * CardUtils.comparePower(Card(Rank.Ace, Suit.Diamond), Card(Rank.Six, Suit.Heart));  // => negative integer
   * CardUtils.comparePower(Card(Rank.Queen, Suit.Spade), Card(Rank.Ace, Suit.Heart));  // => positive integer
   * CardUtils.comparePower(Card(Rank.Ace, Suit.Club), Card(Rank.Ace, Suit.Club));  // => 0
   * ```
   */
  comparePower(a: Card, b: Card): number {
    return (
      RankUtils.comparePower(CardUtils.rankOf(a), CardUtils.rankOf(b)) * 4 +
      SuitUtils.compare(CardUtils.suitOf(a), CardUtils.suitOf(b))
    );
  },

  /**
   * Stringifies a Card.
   *
   * @example
   * ```
   * CardUtils.format(CardUtils.create(Rank.Ace, Suit.Spade)) === ("As" as CardString);
   * CardUtils.format(CardUtils.create(Rank.Deuce, Suit.Spade)) === ("2s" as CardString);
   * CardUtils.format(CardUtils.create(Rank.King, Suit.Club)) === ("Kc" as CardString);
   * ```
   */
  format(card: Card): CardString {
    const rankChar = RankUtils.format(CardUtils.rankOf(card));
    const suitChar = SuitUtils.format(CardUtils.suitOf(card));

    return `${rankChar}${suitChar}`;
  },
});
