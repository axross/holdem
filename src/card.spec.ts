import { Card } from "..";
import { CardUtils } from "./card";
import { Rank } from "./rank";
import { Suit } from "./suit";

describe("CardUtils.create()", () => {
  it("creates Card<As> from Rank.Ace and Suit.Spade", () => {
    expect(CardUtils.create(Rank.Ace, Suit.Spade)).toBe(1);
  });

  it("creates Card<2s> from Rank.Deuce and Suit.Spade", () => {
    expect(CardUtils.create(Rank.Deuce, Suit.Spade)).toBe(2);
  });

  it("creates Card<Ad> from Rank.Ace and Suit.Diamond", () => {
    expect(CardUtils.create(Rank.Ace, Suit.Diamond)).toBe(2 ** 26);
  });

  it("creates Card<Ad> from Rank.King and Suit.Club", () => {
    expect(CardUtils.create(Rank.King, Suit.Club)).toBe(2 ** 51);
  });
});

describe("CardUtils.parse()", () => {
  it('parses "As" into Card<As>', () => {
    expect(CardUtils.parse("As")).toBe(CardUtils.create(Rank.Ace, Suit.Spade));
  });

  it('parses "2s" into Card<2s>', () => {
    expect(CardUtils.parse("2s")).toBe(
      CardUtils.create(Rank.Deuce, Suit.Spade)
    );
  });

  it('parses "Ad" into Card<Ad>', () => {
    expect(CardUtils.parse("Ad")).toBe(
      CardUtils.create(Rank.Ace, Suit.Diamond)
    );
  });

  it('parses "Kc" into Card<Ad>', () => {
    expect(CardUtils.parse("Kc")).toBe(CardUtils.create(Rank.King, Suit.Club));
  });

  it('throws an error because "AS" is invalid string', () => {
    expect(() => CardUtils.parse("AS")).toThrowErrorMatchingSnapshot();
  });

  it('throws an error because "as" is invalid string', () => {
    expect(() => CardUtils.parse("as")).toThrowErrorMatchingSnapshot();
  });

  it('throws an error because "sA" is invalid string', () => {
    expect(() => CardUtils.parse("sA")).toThrowErrorMatchingSnapshot();
  });

  it('throws an error because "Ak" is invalid string', () => {
    expect(() => CardUtils.parse("Ak")).toThrowErrorMatchingSnapshot();
  });

  it('throws an error because "" is invalid string', () => {
    expect(() => CardUtils.parse("")).toThrowErrorMatchingSnapshot();
  });
});

describe("CardUtils.random()", () => {
  it("creates a Card at random", () => {
    const map = new Map<Card, number>();

    for (let i = 0; i < 10000; ++i) {
      const card = CardUtils.random();

      map.set(card, (map.get(card) ?? 0) + 1);
    }

    expect(map.size).toBeGreaterThan(26);
    expect(Math.max(...map.values())).toBeLessThan(1000);
  });
});

describe("CardUtils.rankOf()", () => {
  it("returns Rank.Ace from Card<As>", () => {
    expect(CardUtils.rankOf(CardUtils.create(Rank.Ace, Suit.Spade))).toBe(
      Rank.Ace
    );
  });

  it("returns Rank.Deuce from Card<2c>", () => {
    expect(CardUtils.rankOf(CardUtils.create(Rank.Deuce, Suit.Club))).toBe(
      Rank.Deuce
    );
  });

  it("returns Rank.Nine from Card<9d>", () => {
    expect(CardUtils.rankOf(CardUtils.create(Rank.Nine, Suit.Diamond))).toBe(
      Rank.Nine
    );
  });

  it("returns Rank.Queen from Card<Qh>", () => {
    expect(CardUtils.rankOf(CardUtils.create(Rank.Queen, Suit.Heart))).toBe(
      Rank.Queen
    );
  });
});

describe("CardUtils.suitOf", () => {
  it("returns Suit.Spade from Card<As>", () => {
    expect(CardUtils.suitOf(CardUtils.create(Rank.Ace, Suit.Spade))).toBe(
      Suit.Spade
    );
  });

  it("returns Suit.Club from Card<2c>", () => {
    expect(CardUtils.suitOf(CardUtils.create(Rank.Deuce, Suit.Club))).toBe(
      Suit.Club
    );
  });

  it("returns Suit.Diamond from Card<9d>", () => {
    expect(CardUtils.suitOf(CardUtils.create(Rank.Nine, Suit.Diamond))).toBe(
      Suit.Diamond
    );
  });

  it("returns Suit.Heart from Card<Qh>", () => {
    expect(CardUtils.suitOf(CardUtils.create(Rank.Queen, Suit.Heart))).toBe(
      Suit.Heart
    );
  });
});

describe("CardUtils.compare()", () => {
  it("returns negative integer as comparison result of Card<As> and Card<Ad>", () => {
    expect(
      CardUtils.compare(
        CardUtils.create(Rank.Ace, Suit.Spade),
        CardUtils.create(Rank.Ace, Suit.Diamond)
      )
    ).toBeLessThan(0);
  });

  it("returns positive integer as comparison result of Card<Ad> and Card<6s>", () => {
    expect(
      CardUtils.compare(
        CardUtils.create(Rank.Ace, Suit.Diamond),
        CardUtils.create(Rank.Six, Suit.Heart)
      )
    ).toBeGreaterThan(0);
  });

  it("returns 0 as comparison result of Card<Ac> and Card<Ac>", () => {
    expect(
      CardUtils.compare(
        CardUtils.create(Rank.Ace, Suit.Club),
        CardUtils.create(Rank.Ace, Suit.Club)
      )
    ).toBe(0);
  });
});

describe("CardUtils.comparePower()", () => {
  it("returns negative integer as comparison result of Card<As> and Card<Ad>", () => {
    expect(
      CardUtils.comparePower(
        CardUtils.create(Rank.Ace, Suit.Spade),
        CardUtils.create(Rank.Ace, Suit.Diamond)
      )
    ).toBeLessThan(0);
  });

  it("returns negative integer as comparison result of Card<Ad> and Card<6h>", () => {
    expect(
      CardUtils.comparePower(
        CardUtils.create(Rank.Ace, Suit.Diamond),
        CardUtils.create(Rank.Six, Suit.Heart)
      )
    ).toBeLessThan(0);
  });

  it("returns positive integer as comparison result of Card<Qs> and Card<Ah>", () => {
    expect(
      CardUtils.comparePower(
        CardUtils.create(Rank.Queen, Suit.Spade),
        CardUtils.create(Rank.Ace, Suit.Heart)
      )
    ).toBeGreaterThan(0);
  });

  it("returns 0 as comparison result of Card<Ac> and Card<Ac>", () => {
    expect(
      CardUtils.comparePower(
        CardUtils.create(Rank.Ace, Suit.Club),
        CardUtils.create(Rank.Ace, Suit.Club)
      )
    ).toBe(0);
  });
});

describe("CardUtils.format()", () => {
  it('stringifies Card<As> into "As"', () => {
    expect(CardUtils.format(CardUtils.create(Rank.Ace, Suit.Spade))).toBe(
      "As"
    );
  });

  it('stringifies Card<2c> into "2c"', () => {
    expect(CardUtils.format(CardUtils.create(Rank.Deuce, Suit.Club))).toBe(
      "2c"
    );
  });

  it('stringifies Card<9d> into "9d"', () => {
    expect(CardUtils.format(CardUtils.create(Rank.Nine, Suit.Diamond))).toBe(
      "9d"
    );
  });

  it('stringifies Card<Qh> into "Qh"', () => {
    expect(CardUtils.format(CardUtils.create(Rank.Queen, Suit.Heart))).toBe(
      "Qh"
    );
  });
});
