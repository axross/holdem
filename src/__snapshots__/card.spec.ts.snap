// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardUtils.parse() throws an error because "" is invalid string 1`] = `""" is not a valid string for CardUtils.parse()."`;

exports[`CardUtils.parse() throws an error because "AS" is invalid string 1`] = `""AS" is not a valid string for CardUtils.parse()."`;

exports[`CardUtils.parse() throws an error because "Ak" is invalid string 1`] = `""Ak" is not a valid string for CardUtils.parse()."`;

exports[`CardUtils.parse() throws an error because "as" is invalid string 1`] = `""as" is not a valid string for CardUtils.parse()."`;

exports[`CardUtils.parse() throws an error because "sA" is invalid string 1`] = `""sA" is not a valid string for CardUtils.parse()."`;
